From 01bd49cb03a0dc86cd6702e355f233d7993836a2 Mon Sep 17 00:00:00 2001
From: Christophe Kerello <christophe.kerello@foss.st.com>
Date: Thu, 25 Apr 2024 09:19:32 +0200
Subject: [PATCH 598/841] feat(st-spi): add OSPI support for MP21

On SoCs where 2 OPSI instances are available, OSPI node is found under
OMM node and OSPI parameters are shared between these 2 nodes. It is
the case for STM32MP25 and STM32MP23 SoCs.
There is also SoCs where only one OSPI instance is available. It is
the case for STM32MP21 SoC.

The patch handles these 2 bindings to be able to retrieve OSPI
parameters in the device tree.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 6e29f9fda5cf21f437b3d7317906cb5cdd765f5c feat(st-spi): add OSPI support for MP21]
Change-Id: Ie9f21d7bcff35708b616ff0323d6ed8063b7a3f3
Signed-off-by: Christophe Kerello <christophe.kerello@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/376438
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
Tested-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/st/spi/stm32_ospi.c | 143 +++++++++++++++++++++---------------
 1 file changed, 84 insertions(+), 59 deletions(-)

diff --git a/drivers/st/spi/stm32_ospi.c b/drivers/st/spi/stm32_ospi.c
index 7e5a7c5c8..cbb84aad3 100644
--- a/drivers/st/spi/stm32_ospi.c
+++ b/drivers/st/spi/stm32_ospi.c
@@ -1188,83 +1188,108 @@ int stm32_ospi_init(void)
 
 	iom_node = fdt_node_offset_by_compatible(fdt, -1, _DT_IOM_COMPAT);
 	if (iom_node < 0) {
-		return iom_node;
-	}
+		struct dt_node_info info;
 
-	if (fdt_get_status(iom_node) == DT_DISABLED) {
-		return -FDT_ERR_NOTFOUND;
-	}
+		ospi_node = dt_get_node(&info, -1, _DT_OSPI_COMPAT);
+		if (ospi_node < 0) {
+			return ospi_node;
+		}
 
-	ret = fdt_get_reg_props_by_name(fdt, iom_node, "omm_mm",
-					&mm_base, &mm_size);
-	if (ret != 0) {
-		return ret;
-	}
+		if (info.status == DT_DISABLED) {
+			return -FDT_ERR_NOTFOUND;
+		}
 
-	cuint = fdt_getprop(fdt, iom_node, "ranges", NULL);
-	if (cuint == NULL) {
-		return -FDT_ERR_BADVALUE;
-	}
+		stm32_ospi.reg_base = info.base;
+		stm32_ospi.mm_size = SZ_256M;
+		stm32_ospi.mm_base = STM32MP_OSPI_MM_BASE;
+		stm32_ospi.bank = 0U;
 
-	for (i = 0U; i < _OMM_MAX_OSPI; i++) {
-		bank = fdt32_to_cpu(*cuint);
-		if ((bank >= _OMM_MAX_OSPI) ||
-		    ((bank_assigned & BIT(bank)) != 0U)) {
+		if (info.clock < 0) {
 			return -FDT_ERR_BADVALUE;
 		}
 
-		bank_assigned |= BIT(bank);
-		bank_address[bank] = fdt32_to_cpu(*(cuint + 2U));
-		cuint += 4U;
-	}
+		stm32_ospi.clock_id = (unsigned long)info.clock;
 
-	if (dt_set_pinctrl_config(iom_node) != 0) {
-		return -FDT_ERR_BADVALUE;
-	}
+		if (dt_set_pinctrl_config(ospi_node) != 0) {
+			return -FDT_ERR_BADVALUE;
+		}
+	} else {
+		if (fdt_get_status(iom_node) == DT_DISABLED) {
+			return -FDT_ERR_NOTFOUND;
+		}
 
-	fdt_for_each_subnode(ospi_node, fdt, iom_node) {
-		nb_ospi_nodes++;
-	}
+		ret = fdt_get_reg_props_by_name(fdt, iom_node, "omm_mm",
+						&mm_base, &mm_size);
+		if (ret != 0) {
+			return ret;
+		}
 
-	if (nb_ospi_nodes != 1U) {
-		WARN("Only one OSPI node supported\n");
-		return -FDT_ERR_BADVALUE;
-	}
+		cuint = fdt_getprop(fdt, iom_node, "ranges", NULL);
+		if (cuint == NULL) {
+			return -FDT_ERR_BADVALUE;
+		}
 
-	/* Parse OSPI controller node */
-	ospi_node = fdt_node_offset_by_compatible(fdt, iom_node,
-						  _DT_OSPI_COMPAT);
-	if (ospi_node < 0) {
-		return ospi_node;
-	}
+		for (i = 0U; i < _OMM_MAX_OSPI; i++) {
+			bank = fdt32_to_cpu(*cuint);
+			if ((bank >= _OMM_MAX_OSPI) ||
+			    ((bank_assigned & BIT(bank)) != 0U)) {
+				return -FDT_ERR_BADVALUE;
+			}
 
-	if (fdt_get_status(ospi_node) == DT_DISABLED) {
-		return -FDT_ERR_NOTFOUND;
-	}
+			bank_assigned |= BIT(bank);
+			bank_address[bank] = fdt32_to_cpu(*(cuint + 2U));
+			cuint += 4U;
+		}
 
-	cuint = fdt_getprop(fdt, ospi_node, "reg", NULL);
-	if (cuint == NULL) {
-		return -FDT_ERR_BADVALUE;
-	}
+		if (dt_set_pinctrl_config(iom_node) != 0) {
+			return -FDT_ERR_BADVALUE;
+		}
 
-	bank = fdt32_to_cpu(*cuint);
-	if (bank >= _OMM_MAX_OSPI) {
-		return -FDT_ERR_BADVALUE;
-	}
+		fdt_for_each_subnode(ospi_node, fdt, iom_node) {
+			nb_ospi_nodes++;
+		}
 
-	stm32_ospi.reg_base = fdt32_to_cpu(*(cuint + 1U)) + bank_address[bank];
-	stm32_ospi.mm_size = stm32mp_syscfg_get_mm_size(bank);
-	stm32_ospi.mm_base = bank == 0U ?
-			     mm_base : mm_base + mm_size - stm32_ospi.mm_size;
-	stm32_ospi.bank = bank;
+		if (nb_ospi_nodes != 1U) {
+			WARN("Only one OSPI node supported\n");
+			return -FDT_ERR_BADVALUE;
+		}
+
+		/* Parse OSPI controller node */
+		ospi_node = fdt_node_offset_by_compatible(fdt, iom_node,
+							  _DT_OSPI_COMPAT);
+		if (ospi_node < 0) {
+			return ospi_node;
+		}
 
-	cuint = fdt_getprop(fdt, ospi_node, "clocks", NULL);
-	if (cuint == NULL) {
-		return -FDT_ERR_BADVALUE;
+		if (fdt_get_status(ospi_node) == DT_DISABLED) {
+			return -FDT_ERR_NOTFOUND;
+		}
+
+		cuint = fdt_getprop(fdt, ospi_node, "reg", NULL);
+		if (cuint == NULL) {
+			return -FDT_ERR_BADVALUE;
+		}
+
+		bank = fdt32_to_cpu(*cuint);
+		if (bank >= _OMM_MAX_OSPI) {
+			return -FDT_ERR_BADVALUE;
+		}
+
+		stm32_ospi.reg_base = fdt32_to_cpu(*(cuint + 1U)) + bank_address[bank];
+		stm32_ospi.mm_size = stm32mp_syscfg_get_mm_size(bank);
+		stm32_ospi.mm_base = bank == 0U ?
+				     mm_base : mm_base + mm_size - stm32_ospi.mm_size;
+		stm32_ospi.bank = bank;
+
+		cuint = fdt_getprop(fdt, ospi_node, "clocks", NULL);
+		if (cuint == NULL) {
+			return -FDT_ERR_BADVALUE;
+		}
+
+		cuint++;
+		stm32_ospi.clock_id = (unsigned long)fdt32_to_cpu(*cuint);
 	}
 
-	cuint++;
-	stm32_ospi.clock_id = (unsigned long)fdt32_to_cpu(*cuint);
 	clk_enable(stm32_ospi.clock_id);
 
 	cuint = fdt_getprop(fdt, ospi_node, "resets", &len);
-- 
2.47.2

